src/netaddress.h

```diff
@@ -81,19 +116,61 @@ class CNetAddr
         bool GetInAddr(struct in_addr* pipv4Addr) const;
         std::vector<unsigned char> GetGroup() const;
         int GetReachabilityFrom(const CNetAddr *paddrPartner = nullptr) const;
-
+        bool GetV1Serialization(unsigned char *buff) const;
+        void FromV1Serialization(unsigned char *buff);
         explicit CNetAddr(const struct in6_addr& pipv6Addr, const uint32_t scope = 0);
         bool GetIn6Addr(struct in6_addr* pipv6Addr) const;
 
         friend bool operator==(const CNetAddr& a, const CNetAddr& b);
         friend bool operator!=(const CNetAddr& a, const CNetAddr& b) { return !(a == b); }
         friend bool operator<(const CNetAddr& a, const CNetAddr& b);
 
-        ADD_SERIALIZE_METHODS;
+        template<typename Stream>
+        void Serialize(Stream &s) const
+        {
+            if (s.GetVersion() & SERIALIZE_ADDR_AS_V2) {
+                s << static_cast<uint8_t>(m_network_id);
+                s << ip;
+            } else {
+                unsigned char ip_temp[16];
+                GetV1Serialization(ip_temp);
```

Comment:
We must make sure that we never end up here with `this` being one of the new address types, e.g. Tor v3. If that happens `GetV1Serialization()` will return `false` (which is currently ignored).

We should not attempt to send e.g. Tor v3 address to a peer that did not signal ADDRv2 support. Maybe `CNode::PushAddress()` is a good place to plant a check for that and skip adding new type addresses to the send queue of an old peer.